### ISSUE TITLE
Improved Image Handling

### ISSUE THREAD

**Issue by dullage** [dullage/flatnotes#45]
Currently, images are inserted as inline base64 which isn't ideal. Equally, other attachment types (e.g. PDFs) are not supported at all. It would be good to improve this.

**Comment by lmingzhi** [comment]
Your project is great! I'm currently using it to display my personal notes on a server host, so that I can search them anywhere through the web (via a VPN to my inner network).

Since Flatnotes does not support folder hierarchies, I can accept that, as I use `Logseq` to edit my notes and simply use `rsync` to sync the notes to the remote server.

However, one inconvenience is that Flatnotes cannot recognize my image files, which are stored in a local folder called assets, like `![image.png](../assets/image_1695872231710_0.png)`.

I would hope that Flatnotes could support local images in Markdown files in the future.

I really appreciate your work on this project. Thank you again!

**Comment by lmingzhi** [comment]
I create a temporary solution.

```
# Goto my host flatnotes path
cd /nvme/38080_flatnotes
git clone https://github.com/dullage/flatnotes.git

root@dt120 ➜  38080_flatnotes ll
total 152K
drwxr-xr-x 2 lmingzhi lmingzhi  72K Sep 28 11:53 assets
drwxr-xr-x 3 lmingzhi lmingzhi  68K Sep 28 09:10 data
-rw-r--r-- 1 lmingzhi root      669 Sep 28 14:51 docker-compose.yml
drwxr-xr-x 6 root     root     4.0K Sep 28 14:47 flatnotes
drwxr-xr-x 2 lmingzhi lmingzhi 4.0K Sep  6 18:51 index
```

the  `docker-compose.yml` file:

```
version: "3"

# https://github.com/Dullage/flatnotes#self-hosted
services:
  flatnotes:
    container_name: flatnotes
    image: dullage/flatnotes:latest
    environment:
      PUID: 1001
      PGID: 1001
      FLATNOTES_AUTH_TYPE: "password"
      FLATNOTES_USERNAME: "lmingzhi"
      FLATNOTES_PASSWORD: "**YOUR PERSONAL SECREAT PASSWORD**"
      FLATNOTES_SECRET_KEY: "HHHHaLongRandomSeriesOfCharacters"
    volumes:
      - "./data:/data"
      - "./assets:/assets" 
      - "./flatnotes/flatnotes:/app/flatnotes"
      # Optional. Allows you to save the search index in a different location:
      # - "./index:/data/.flatnotes"
    ports:
      - "38080:8080"
    restart: unless-stopped
```

The attach is store in `assets` folder, as I mention before that my image link looks like `![image.png](../assets/image_1695872231710_0.png)`.

Do following change:

1. copy `dist` folder from docker to local flatnotes path

```bash
docker exec -it flatnotes bash
# In docker 
root@129762581a05:/app/flatnotes# cp -r /app/flatnotes/dist/ /data
# In Host
root@dt120 ➜ mv data/dist flatnotes/flatnotes
```

2. modify code

```python
# vi flatnotes/flatnotes/main.py
# Add follow code:
# https://stackoverflow.com/questions/70740267/insert-local-image-in-the-fastapi-automatic-documentation

from fastapi.responses import FileResponse
import os
# An endpoint to serve images for documentation
@app.get("/assets/{filename}")
def get_img(filename: str):
    filepath = os.path.join('/assets/', os.path.basename(filename))
    return FileResponse(filepath)
```

3. Restart the container `docker-compose restart`

----

logs:

```bash
# In beginning
flatnotes    | INFO:     192.168.100.117:64551 - "GET /assets/image_1695872231710_0.png HTTP/1.1" 404 Not Found
flatnotes    | INFO:     192.168.100.117:64567 - "GET /assets/image_1695873069248_0.png HTTP/1.1" 404 Not Found

# After
flatnotes    | Setting up user and group...
flatnotes    | addgroup: The group `flatnotes' already exists.
flatnotes    | Group '1001' already exists.
flatnotes    | User '1001' already exists.
flatnotes    | Setting file permissions...
flatnotes    | adduser: The user `flatnotes' already exists.
flatnotes    | WARNING: Breaking changes introduced in version 3.x:
flatnotes    |   - The port flatnotes uses inside the Docker container has been changed to 8080 (previously 80).
flatnotes    |   - To accompany the above change, support for the PORT environment variable has been removed.
flatnotes    |   - The note directory inside the Docker container has moved from /app/data to simply /data.
flatnotes    | Starting flatnotes...
flatnotes    | 2023-09-28 06:55:51 [INFO]: Loading existing index
flatnotes    | INFO:     Started server process [1]
flatnotes    | INFO:     Waiting for application startup.
flatnotes    | INFO:     Application startup complete.
flatnotes    | INFO:     Uvicorn running on http://0.0.0.0:8080 (Press CTRL+C to quit)
flatnotes    | INFO:     192.168.100.117:65378 - "GET /note/demo-notes HTTP/1.1" 200 OK
flatnotes    | INFO:     192.168.100.117:65378 - "GET /api/config HTTP/1.1" 200 OK
flatnotes    | INFO:     192.168.100.117:65379 - "GET /assets/image_1695872231710_0.png HTTP/1.1" 200 OK
flatnotes    | INFO:     192.168.100.117:65378 - "GET /assets/image_1695873069248_0.png HTTP/1.1" 200 OK

```

**Comment by dullage** [comment]
This feature is now available in the `develop` branch. If anyone is interested in giving it a go before it is officially released they can use the Docker tag `:develop`. It would be great to get some early beta testers.

## Draft Wiki Article

There are a number of ways to add images to a note:

- Using the 'Insert image' button in the editor toolbar.
- Dragging and dropping an image file into the editor.
- Copying and pasting an image into the editor.

These images are then uploaded to a folder named "attachments" located within your flatnotes data folder. Note: If this folder doesn't exist, it is created on startup.

All files in the attachments folder are accessible at `/attachments/<filename>`. For example, if you have an image named `my-image.png` in the attachments folder, you can access it at `https://www.your-flatnotes-url.com/attachments/my-image.png`.

Currently, there is no way to view or delete files in the attachments folder from flatnotes and they are not deleted when the note is deleted. You will need to manually delete them from the folder as required.

**Comment by OPerepadia** [comment]
I just tried it, and its working perfectly. Thanks for your work on this project!

**Comment by dullage** [comment]
Great to hear, thanks for the feedback.

**Comment by dullage** [comment]
Image support has now been released in version v3.4.0. I have created a separate suggestion for generic attachment support, see #134.
